JMMServer/Tasks/AutoAnimeGroupCalculator.cs

```diff
@@ -0,0 +1,594 @@
+            {
+                return false;
+            }
+            if (rel.RelationType != AnimeRelationType.AlternativeSetting && rel.RelationType != AnimeRelationType.Other
+                && rel.RelationType != AnimeRelationType.SameSetting)
+            {
+                return true;
+            }
+
+            // Perform a very poor man's string metric test.
+            // We split both titles up into tokens (basically words) and count the number of times each word
+            // appears in both titles (we'll also count characters of matching words).
+            // The more words/characters that match, the more likely we'll consider the relationship
+            string[] fromTitleTokens = CreateTokensFromTitle(rel.FromMainTitle);
+            string[] toTitleTokens = CreateTokensFromTitle(rel.ToMainTitle);
+            int matchLen = 0;
+            int matches = 0;
+
+            foreach (string fromToken in fromTitleTokens)
+            {
+                foreach (string toToken in toTitleTokens)
+                {
+                    if (fromToken.Equals(toToken, StringComparison.InvariantCultureIgnoreCase))
+                    {
+                        matchLen += fromToken.Length;
+                        matches++;
+                        break;
+                    }
+                }
+            }
+
+            if (matches == 0)
+            {
+                return false;
+            }
+
+            int shortestTitleLen = Math.Min(fromTitleTokens.Sum(s => s.Length), toTitleTokens.Sum(s => s.Length));
+            int minTokenCount = Math.Min(fromTitleTokens.Length, toTitleTokens.Length);
+
+            // Either approximately half the words must match,
+            // or the total length of the matched words must equate to 40% or more of the shortest title
+            return (matches >= minTokenCount / 2) || matchLen >= Math.Max(1, (int)(shortestTitleLen * 0.4));
+        }
+
+        private static string[] CreateTokensFromTitle(string title)
+        {
+            title = title.Replace('-', ' ');
+            title = TitleNoiseRegex.Replace(title, String.Empty);
+
+            return title.Split(Whitespace, StringSplitOptions.RemoveEmptyEntries);
+        }
+
+        public AutoGroupExclude Exclusions
+        {
+            get { return _exclusions; }
+        }
+
+        [DebuggerDisplay("{AnimeNode.AnimeId} (Score: {Score})")]
+        private sealed class AnimeRelationStats
+        {
+            public RelationNode AnimeNode;
+            public int Sequels;
+            public int AlternateVersions;
+            public int FullStory;
+            public int ParentStory;
+
+            /// <summary>
+            /// Gets the calculated score for a particular anime.
+            /// The higher the score, the higher the priority this anime should be given.
+            /// </summary>
+            public int Score

```

Comment:
This is quite interesting, and I'll need to play with it. I'll need to Make a setting for those who purely want the first aired to be the main, though. I'll also need to look into using said setting and have the desired style for all clients as well.